### ISSUE TITLE
Worlds: ChunkUtils (or WorldUtils)

### ISSUE THREAD

**Issue by Bavfalcon9** [JSPrismarine/JSPrismarine#296]
When multiple players on on a server, loading chunks at a fast rate, TPS significantly drops. This issue is to discuss alternatives.

**Alternative 1:** 
I haven't done any bench marking tests for this, but the general idea is that each world generator that takes more than 10ms to generate a single chunk should be moved to WASM. I personally don't prefer this, because there's a lot more that can be handled in wasm, which I'll describe in alternative 2. However, the general concept for this is that instead of something like:
```ts
for (let x = 0; x < 16; x++) {
     for (let z = 0; z < 16; z++) {
          someCPUIntensiveTask();
          modifyChunksWithIntensiveTask(); // eg: PerloinNoise
     }
}
```
We have,
````ts
const chunkData: BlockData[] = gen_overworld_chunk();
chunkData.forEach(block => { chunk.setBlock(block.x, block.y, block.z, block.name) });
```

**Alternative 2:**
Because of how CPU Intensive generating chunks, reading and writing from system IO is, (level db, etc) we can implement the entire level system in WASM, or a extension for NodeJS and provide bindings for it. The reason behind this is so we can get the full benefits of using WASM or a lower level language. Bindings to this utility would be something like:
```ts
import * as WorldUtils from 'worldOrChunkutils';

function onChunkPublish((chunk) => {
 // send chunk to nearby players
});
const level = WorldUtils.getLevel('someName', 'someGenerator');

// load or get chunk
level.getChunk(x, z);
```

**Comment by filfat** [comment]
I prefer the 1st alternative even if it’s quite a bit slower. Allowing plugin authors to create their own providers and generators is rather important. Eg for Skyblock or similar servers.

**Comment by Bavfalcon9** [comment]
You would still be able to do it with the second alternative, there would just be an api for it

**Comment by filfat** [comment]
> You would still be able to do it with the second alternative, there would just be an api for it.
> I still need to brainstorm how this api would look though.

You’re right and wasm would be possible since it would be up to the provider to handle fs (luckily node fs is extremely fast since it’s all native-ish).

**Comment by Bavfalcon9** [comment]
> 
> 
> > You would still be able to do it with the second alternative, there would just be an api for it.
> > I still need to brainstorm how this api would look though.
> 
> You’re right and wasm would be possible since it would be up to the provider to handle fs (luckily node fs is extremely fast since it’s all native-ish).

Alright, I'll start writing out some API methods for this, and get to work on it. 👍 
cc: @HerryYT
